native/src/main/java/io/ballerina/lib/avro/visitor/DeserializeVisitor.java

```diff
@@ -0,0 +1,411 @@
+                        Object[] objectArray = ((GenericData.Array<?>) fieldData).toArray();
+                        if (schemaType.getElementType().getType().equals(Schema.Type.STRING)
+                                || schemaType.getElementType().getType().equals(Schema.Type.ENUM)) {
+                            BString[] stringArray = new BString[objectArray.length];
+                            BArray ballerinaArray = ValueCreator.createArrayValue(stringArray);
+                            int i = 0;
+                            for (Object obj : objectArray) {
+                                stringArray[i] = StringUtils.fromString(obj.toString());
+                                i++;
+                            }
+                            avroRecord.put(StringUtils.fromString(field.name()), ballerinaArray);
+                        } else {
+                            avroRecord.put(StringUtils.fromString(field.name()), fieldData);
+                        }
+                    }
+                }
+                case MAP -> {
+                    if (fieldData instanceof Map<?, ?>) {
+                        BMap<BString, Object> avroMap = ValueCreator.createMapValue();
+                        Object[] keys = ((Map<String, Object>) fieldData).keySet().toArray();
+                        for (Object key : keys) {
+                            avroMap.put(StringUtils.fromString(key.toString()),
+                                    ((Map<String, Object>) fieldData).get(key));
+                        }
+                        avroRecord.put(StringUtils.fromString(field.name()), avroMap);
+                    }
+                }
+                case RECORD -> {
+                    if (fieldData instanceof GenericRecord) {
+                        avroRecord.put(StringUtils.fromString(field.name()),
+                                       visitRecords(type, schemaType, (GenericRecord) fieldData));
+                    }
+                }
+                case STRING -> {
+                    if (fieldData instanceof Utf8) {
+                        avroRecord.put(StringUtils.fromString(field.name()),
+                                       StringUtils.fromString(fieldData.toString()));
+                    }
+                }
+                case INT, LONG -> {
+                    if (fieldData instanceof Integer || fieldData instanceof Long) {
+                        avroRecord.put(StringUtils.fromString(field.name()),
+                                       ((Number) fieldData).longValue());
+                    }
+                }
+                case FLOAT, DOUBLE -> {
+                    if (fieldData instanceof Double) {
+                        avroRecord.put(StringUtils.fromString(field.name()), fieldData);
+                    } else {
+                        avroRecord.put(StringUtils.fromString(field.name()), Double.parseDouble(fieldData.toString()));
```

Comment:
This function is too long and has a higher cognitive complexity. Let's try to break it into smaller functions. The loops and if conditions inside switch statements is a good place to start to.